Generate Logging

michaelbausor opened the pull request:
This PR separates the generation of Logging from Pubsub in https://github.com/GoogleCloudPlatform/gcloud-php/pull/62

One point for discussion/review is that the json files with retry settings are in a top level directory called resources. What would be the best way to provide these files?

jdpedrie commented on phpunit.xml.dist:
```diff
@@ -8,6 +8,9 @@
     <whitelist>
       <directory suffix=".php">src</directory>
+      <exclude>
+        <directory suffix=".php">src/Logging/V2</directory>
```
What would everyone think of using globs here to cover these without needing to add more excludes for every new API surface?

`src/*/V*`, or `src/*/V1`, `src/*/V2` should cover it.

garrettjonesgoogle replied:
We would want V3 also (for Monitoring)

michaelbausor replied:
How about:
`src/*/V?`

garrettjonesgoogle replied:
Do we ever want to support beta versions of APIs, e.g. v1beta2?

dwsupplee replied:
That seems like something we should look out for, yes.

`src/*/V?` and `src/*/V?beta?` then?

Should we plan for cases where a service reaches V10+?

jdpedrie replied:
I played around with the globs a bit, and this seems to work. It should be forward-compatible with newer API versions and potential subfolders beginning with "V". It should match any folder beginning with "V", immediately followed by a non-apha character.

````xml
  <filter>
    <whitelist>
      <directory suffix=".php">src</directory>
      <exclude>
        <directory suffix=".php">src/*/V[!a-zA-Z]</directory>
      </exclude>
    </whitelist>
  </filter>
````

garrettjonesgoogle replied:
That definitely seems like the most robust logic - does it also handle the `v1beta2` case?

dwsupplee replied:
`src/*/V[!a-zA-Z]*` should cover that

michaelbausor replied:
Looks good. We can exclude V10+ and betas with a trailing star:
`<directory suffix=".php">src/*/V[0-9]*</directory>`

michaelbausor replied:
Oops, didn't see the previous two comments, before posting mine

garrettjonesgoogle commented:
Is there anything else that needs to be addressed on this PR before it can be merged?

dwsupplee commented:
This is looking good. The only issue on this PR I would like to discuss still is in regards to @michaelbausor's point regarding the resources directory.

We haven't decided whether we will publish each API as it's own package yet. If we do, however, placing the resource files in src/Logging would help make the transition easier. What do you think?

michaelbausor commented:
Does it make sense to place them inside the major version folder? e.g. `src/Logging/V2/resources` That way it will be easier to manage API changes between major versions that cause the retry settings to change.

dwsupplee commented:
That sounds perfect!

dwsupplee commented on src/Logging/V2/ConfigServiceV2Api.php:
```diff
@@ -0,0 +1,601 @@
+
+        // TODO load the client config in a more package-friendly way
+        // https://github.com/googleapis/toolkit/issues/332
+        $clientConfigJsonString = file_get_contents('./resources/config_service_v2_client_config.json');
```
We will need to update these to something like:

`file_get_contents(__DIR__ . '/resources/config_service_v2_client_config.json');`

bshaffer commented:
Why do the classes end in `V2API`? The `V2` namespace is enough to prevent collisions.

michaelbausor commented:
We are appending the `Api` portion - the other parts are coming from the logging proto, which is not following the style guide. So for example in pubsub they have named their interface `Publisher`, and we generate classes named `PublisherApi`. Instead of just calling their interfaces `Logging`, `Metrics` and `Config`, the interfaces are named `LoggingServiceV2`, `MetricsServiceV2` and `ConfigServiceV2`, so we generate `LoggingServiceV2Api`, etc.

michaelbausor commented:
Sorry I have let this sit for a week. I have updated the logging surface with our latest generator. I also updated the google/auth dependency to 0.10, which is required by google/gax to use the caching functionality that has been added.

Please let me know if there are more issues that need to be addressed.

dwsupplee commented:
Thank you for the updates. 

Should https://github.com/GoogleCloudPlatform/google-cloud-php/pull/62 lead this? It looks like it has the requires for the necessary dependencies. On that note, though, what do you think about moving those dependencies from `require` to `suggest` since we can't rely on a user having the gRPC extension installed?

michaelbausor commented:
That seems sensible to me. @garrettjonesgoogle what are your thoughts?

This PR should probably not block on #62, because the PubSub surface does not yet support IAM methods. Support for those should be coming soon, but isn't available yet.

I will add the dependencies to this PR. Note also that since Stanley has flagged release versions for grpc (https://packagist.org/packages/grpc/grpc#v1.0.0) and protobuf (https://packagist.org/packages/stanley-cheung/protobuf-php#v0.6) we should not require `"minimum-stability": "dev"`

garrettjonesgoogle commented:
@michaelbausor @dwsupplee Moving the dependencies from `require` to `suggest` sounds like a good idea.
